### PR TITLE
[SPARK-47440][SQL] Fix pushing unsupported syntax to MsSqlServer

### PR DESCRIPTION
**Pull request by stefanbuk-db**
### What changes were proposed in this pull request?
In this PR, I propose a change in SQLQuery builder of MsSqlServer dialect. I override build method to check for boolean operator in binary comparisons and throw exception if encountered.

### Why are the changes needed?
MsSqlServer syntax prevents boolean operators in any binary comparison. Reasoning is lack of boolean data type in MsSqlServer.
It was possible to construct Spark query that would generate this situation in MsSqlServer and engine would throw syntax exception on the MsSqlServer side. This PR solves this bug.
For example, in table `people` there is a `name` column. In MsSqlServer if we try to execute:
`SELECT * FROM people WHERE (name LIKE 'a%') = (name LIKE '%b')`
we would get a syntax error. However this query is fine in other major engines.

### Does this PR introduce _any_ user-facing change?
Yes, user will not encounter syntax exception in MsSqlServer when writing these queries.

### How was this patch tested?
By running a unit test in MsSqlServerIntegrationSuite

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for making a PR, @stefanbuk-db . Is this MSSQLServer specific issue?

**Comment by dongjoon-hyun** [comment]
cc @yaooqinn , too

**Comment by stefanbuk-db** [comment]
> Thank you for making a PR, @stefanbuk-db . Is this MSSQLServer specific issue?

Yes, as far as I know this is a MSSQLServer specific issue, not an overall one.

File: common/utils/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -2648,6 +2648,11 @@
           "Cannot resolve window reference <windowName>."
         ]
       },
+      "UNSUPPORTED_BINARY_COMPARISON_LIKE_OPERATOR" : {
```
Comment (yaooqinn) [inline common/utils/src/main/resources/error/error-classes.json:2651]: can we also have a test for the error case?

**Comment by yaooqinn** [comment]
> MsSqlServer syntax prevents LIKE operator in any binary comparison.

Can we have a reference and some comments both in the PR desc and related code changes?

> Yes, user will not encounter syntax exception in MsSqlServer when writing these queries.

Can we make some examples for `these queries` in the PR desc?

**Review by milastdbx — APPROVED** [review]
LGTM, thanks for working on this change!

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala [hunk 2]
```diff
@@ -125,4 +125,10 @@ class MsSqlServerIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JD
       },
       errorClass = "_LEGACY_ERROR_TEMP_2271")
   }
+
+  test("SPARK-47440: SQLServer does not support LIKE in binary comparison") {
+    val df = sql(s"SELECT name FROM " +
+      s"$catalogName.employee WHERE ((name LIKE 'am%') = (name LIKE '%y'))")
```
Comment (yaooqinn) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:131]: `WHERE ((name NOT LIKE 'am%') = (name NOT LIKE '%y'))` is not covered, right?

Comment (stefanbuk-db) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:131]: Yes, that is correct. LIKE is not the only thing that can cause problems with MsSqlServer syntax. In this case that would be operator NOT that also has a boolean return value, which is also a case for other logical operators. That all should be a part of a bigger fix.

Comment (yaooqinn) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:131]: We can fix it together

Comment (stefanbuk-db) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:131]: Sure! What are next steps?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala [hunk 3]
```diff
@@ -86,6 +86,35 @@ private object MsSqlServerDialect extends JdbcDialect {
...
+          }
+        case _ => false
+      }
+
+      // MsSqlServer does not support boolean comparison using standard comparison operators
+      // LIKE operator has a boolean result, hence we don't propagate these queries to MsSqlServer
+      expr match {
+        case e: GeneralScalarExpression =>
+          e.name() match {
+            case "=" | "<>" | "<=>" | "<" | "<=" | ">" | ">=" =>
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala:105]: Please overwrite `visitBinaryComparison`

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala (hunk 2 above)
Comment (yaooqinn) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:131]: let's add the above case to the test and make it work

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala [hunk 4]
```diff
@@ -86,6 +86,35 @@ private object MsSqlServerDialect extends JdbcDialect {
       case "STDDEV_SAMP" => "STDEV"
       case _ => super.dialectFunctionName(funcName)
     }
+
+    override def build(expr: Expression): String = {
```
Comment (beliefer) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala:90]: Please don't override `build` which specify the specification of DS V2 framework.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala (hunk 3 above)
Comment (stefanbuk-db) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala:106]: Problem with overwriting `visitBinaryComparison` is that it takes strings as parameters. This leaves us with not much information about child expressions. Parsing these strings could be very complex and costly, so overriding this function is not the best option here.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala (hunk 4 above)
Comment (stefanbuk-db) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala:90]: See comment above.

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala (hunk 2 above)
Comment (stefanbuk-db) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala]: Added support for multiple logical operators, also added NOT test case that now passes.

**Review by milastdbx — APPROVED** [review]
LGTM, thanks for fixing this!

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala (hunk 3 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala:106]: Shall we overwrite `visitBinaryComparison` and convert the boolean to integer type ? we can avoid the issue.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala [hunk 5]
```diff
@@ -86,6 +87,31 @@ private object MsSqlServerDialect extends JdbcDialect {
...
+                super.visitUnexpectedExpr(expr)
+              } else {
+                super.build(expr)
+              }
+            case _ => super.build(expr)
+          }
+        case _ =>
+          super.build(expr)
+      }
+    }
```
Comment (yaooqinn) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala:114]:
```suggestion
    override def build(expr: Expression): String = {
      // MsSqlServer does not support boolean comparison using standard comparison operators
      // We shouldn't propagate these queries to MsSqlServer
      expr match {
        case e: Predicate => e.name() match {
          case "=" | "<>" | "<=>" | "<" | "<=" | ">" | ">="
              if e.children().exists(_.isInstanceOf[Predicate]) =>
            super.visitUnexpectedExpr(expr)
          case _ => super.build(expr)
        }
        case _ => super.build(expr)
      }
    }
```

Comment (stefanbuk-db) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala]: Thanks for suggestion!

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala [hunk 6]
```diff
@@ -125,4 +125,20 @@ class MsSqlServerIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JD
       },
       errorClass = "_LEGACY_ERROR_TEMP_2271")
   }
+
+  test("SPARK-47440: SQLServer does not support boolean expression in binary comparison") {
+    val df1 = sql(s"SELECT name FROM " +
```
Comment (yaooqinn) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:130]:
```suggestion
    val df1 = sql("SELECT name FROM " +
```

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala [hunk 7]
```diff
@@ -125,4 +125,20 @@ class MsSqlServerIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JD
...
+      s"$catalogName.employee WHERE ((name LIKE 'am%') = (name LIKE '%y'))")
+    assert(df1.collect().length == 4)
+
+    val df2 = sql(s"SELECT name FROM " +
+      s"$catalogName.employee " +
+      s"WHERE ((name NOT LIKE 'am%') = (name NOT LIKE '%y'))")
```
Comment (yaooqinn) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:136]:
```suggestion
      "WHERE ((name NOT LIKE 'am%') = (name NOT LIKE '%y'))")
```

Comment (yaooqinn) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:136]:
```suggestion
      "WHERE ((name NOT LIKE 'am%') = (name NOT LIKE '%y'))")
```

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala [hunk 8]
```diff
@@ -125,4 +125,20 @@ class MsSqlServerIntegrationSuite extends DockerJDBCIntegrationV2Suite with V2JD
...
+    assert(df2.collect().length == 4)
+
+    val df3 = sql(s"SELECT name FROM " +
```
Comment (yaooqinn) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:139]:
```suggestion
    val df3 = sql("SELECT name FROM " +
```

**Review by yaooqinn — APPROVED** [review]
LGTM, only some super nit comments

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala [hunk 9]
```diff
@@ -86,6 +87,20 @@ private object MsSqlServerDialect extends JdbcDialect {
...
     }
+
+    override def build(expr: Expression): String = {
+      // MsSqlServer does not support boolean comparison using standard comparison operators
+      // We shouldn't propagate these queries to MsSqlServer
+      expr match {
+        case e: Predicate => e.name() match {
+          case "=" | "<>" | "<=>" | "<" | "<=" | ">" | ">="
+            if e.children().exists(_.isInstanceOf[Predicate]) =>
+              super.visitUnexpectedExpr(expr)
```
Comment (yaooqinn) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala:98]:
```suggestion
              if e.children().exists(_.isInstanceOf[Predicate]) =>
            super.visitUnexpectedExpr(expr)
```

**Comment by yaooqinn** [comment]
Thank you @stefanbuk-db @beliefer @dongjoon-hyun @milastdbx 

Merged to master/3.5.2

**Comment by dongjoon-hyun** [comment]
Thank you all!

**Pull request closed without merging**